### PR TITLE
Add metrics

### PR DESCRIPTION
**Pull request by eccles**
Metrics package

[AB#8684](https://dev.azure.com/jitsuin/0629f48c-3979-4bbc-9026-cb06b3dfd0ae/_workitems/edit/8684)

### CODE REVIEW

File: metrics/handler.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	if m == nil {
+		return h
+	}
+	m.log.Debugf("NewLatencyMetricsHandler")
+	observer := NewLatencyObservers(m)
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		log := m.log.FromContext(r.Context())
+		defer log.Close()
+
+		fields := strings.Split(strings.Trim(r.URL.Path, "/ "), "/")
```
Comment (honourfish) [inline metrics/handler.go:34]: does the url package give us this without having to do strings.Split and strings.Trim?

File: metrics/handler.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		log.Debugf("Fields: %v (%d)", fields, len(fields))
+		if fields[0] != URLPrefix {
+			h.ServeHTTP(w, r)
+			return
+		}
+		log.Debugf("Request URL Path: '%s'", r.URL.Path)
+		// log.Debugf("Headers before processing: %v", r.Header)
```
Comment (honourfish) [inline metrics/handler.go:41]: minor: commented out line?

File: metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+		prometheus.HistogramOpts{
+			Name:    "archivist_requests_latency",
+			Help:    "Histogram of time to reply to request.",
+			Buckets: []float64{.005, .01, .02, .04, .08, .16, .32},
+		},
+		[]string{"method", "tenant", "service", "resource"},
+	)
+}
+
+// create metric according to proof mechanism of ledger or tenant_storage
```
Comment (honourfish) [inline metrics/metrics.go:66]: nice comment!

File: metrics/observers.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	requestsCounter *prometheus.CounterVec
+	requestsLatency *prometheus.HistogramVec
+	serviceName     string
+	labels          []latencyObserveOffset
+	log             Logger
+}
+
+type LatencyOption func(*LatencyObservers)
+
+//func WithLabel(label string, offset int) LatencyOption {
```
Comment (honourfish) [inline metrics/observers.go:26]: commented out code?

File: tenantid/tenantID.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+	"net/http"
+)
+
+const (
+	grpcMetadataHeaderKey = "grpc-metadata-archivist-internal-tenant-id"
+	headerKey             = "archivist-internal-tenant-id"
+	Prefix                = "tenant/"
+)
+
+func GetTenantIDFromHeader(h http.Header) string {
```
Comment (honourfish) [inline tenantid/tenantID.go:13]: How does this correspond to the functions in wal, particularly tenantid.go and middlware.go.

I think there is also some tenantID logic from headers in idtoken?

Comment (eccles) [reply tenantid/tenantID.go:13]: yes - eventually most will be moved out of wal package and into datatrails-common but not yet.
This PR is the minumum required to move metrics.go from wal to datatrails-common

File: metrics/handler.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+		header := lrw.Header()
+		log.Debugf("Headers after processing: %v", header)
+
+		statusCode := strconv.Itoa(lrw.StatusCode)
+		log.Debugf("status code: %s", statusCode)
+
+		// generate post-process metrics here...
+
+		// tenantID
+		// TODO: when we put the tenant ID in the JWT we can get rid of these
```
[the remark this answers is not in the document]
Comment (eccles) [reply metrics/handler.go:67]: ah - I did not know that

[the remark this answers is not in the document]
Comment (eccles) [reply metrics/handler.go:67]: can you give me more detail

File: metrics/handler.go (hunk 2 above)
Comment (eccles) [reply metrics/handler.go:41]: fixed

File: metrics/handler.go (hunk 1 above)
Comment (eccles) [reply metrics/handler.go:34]: I dont think so - once you have the Path thendoing a split seems to be the accepted method....

File: metrics/observers.go (hunk 4 above)
Comment (eccles) [reply metrics/observers.go:26]: fixeed

File: tenantid/tenantID.go (hunk 5 above)
Comment (honourfish) [reply tenantid/tenantID.go:13]: Can we unit test these then?

File: metrics/handler.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (eccles) [reply metrics/handler.go:66]: yes - but this is out of scope for this PR. I have rasied a ticket to cover this work.

File: tenantid/tenantID.go (hunk 5 above)
Comment (eccles) [reply tenantid/tenantID.go:13]: I would rather do that in separate PR/ticket...

Comment (eccles) [reply tenantid/tenantID.go:13]: ticket raised for doing unittests on this repo

**Review by honourfish — APPROVED** [review]

File: tenantid/tenantID.go (hunk 5 above)
Comment (honourfish) [reply tenantid/tenantID.go:13]: perfect!

**Pull request merged**